smolyay/samples.py

```diff
@@ -0,0 +1,362 @@
+import abc
+import numpy
+
+
+class UnidimensionalPointSet(abc.ABC):
+    """Set of unidimensional points
+
+    A set of unique unidimensional points to be used as the sampling points
+    in a tensor product grid or sparse grid. These unidimensional points are
+    associated with certain polynomial families, and the combination of
+    unique points and polynomial families are the basis for different types
+    of quadrature rules that are used in numerical integration and in
+    approximation.
+    The natural domain of the points specified by :param:``domain`` is
+    the domain in which all the possible points in the set lies within.
+    ``points`` is the list of points that the UnidimensionalPointSet
+    is expected to store.
+    ``domain`` is the domain of the points generated by the point set.
+    """
+
+    def __init__(self):
+        self._points = None
+        self._valid_cache = True
+
+    @property
+    @abc.abstractmethod
+    def domain(self):
+        """list: Domain the sample points come from."""
+        pass
+
+    @property
+    def points(self):
+        """list: Points stored by the set."""
+        if self._valid_cache:
+            self._create()
+        self._valid_cache = False
+        return self._points
+
+    @abc.abstractmethod
+    def _create(self):
+        r"""Generating the points
+
+        An abstract method for generating the points stored by the point set
+        and other parameters.
+        """
+        pass
+
+
+class NestedUnidimensionalPointSet(UnidimensionalPointSet):
+    """Set of unidimensional points with levels
+
+    A set of unique unidimensional points to be used as the sampling points
+    in a tensor product grid or sparse grid. These unidimensional points are
+    associated with certain polynomial families, and the combination of
+    unique points and polynomial families are the basis for different types
+    of quadrature rules that are used in numerical integration and in
+    approximation.
+
+    The number of unique points is controlled by `num_points`, which describes
+    the amount of points added by each level.
+    `points` describes the unique points added as the number of levels increase,
+    and the order of elements is reflective of the level each element first
+    appears in. Some point set are nested, meaning that the points in a previous
+    level reappear in following levels. However, `points` does not attempt to
+    capture this behavior and should be free of duplicates.
+
+    Parameters
+    ----------
+    max_level ; int
+        the maximum level the points are used for
+    """
+
+    def __init__(self, max_level):
+        super().__init__()
+        self.max_level = max_level
+        self._num_points = None
+        self._start_level = None
+        self._end_level = None
```

Comment:
You will want to reorder like this so that the guard below works.
```suggestion
        self._max_level = None
        self._num_points = None
        self._start_level = None
        self._end_level = None

        self.max_level = max_level
```